[Excel] (Custom functions) Troubleshooting article (#806)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-troubleshooting.md

---
ms.date: 03/06/2019
description: Troubleshoot common problems in Excel custom functions.
title: Troubleshoot custom functions (preview)
localization_priority: Priority
---
# Troubleshoot custom functions

When developing custom functions, you may encounter errors in the product while creating and testing your functions.

To resolve issues, you can [enable runtime logging to capture errors](#enable-runtime-logging) and refer to [Excel's native error messages](#check-for-excel-error-messages). Also, check for common mistakes such as not [verifying ssl certificates](#verify-ssl-certificates) properly, [leaving promises unresolved](#ensure-promises-return), and forgetting to [associate your functions](#associate-your-functions).

## Enable runtime logging

> If you are testing your add-in in Office on Windows, you should [enable runtime logging](https://docs.microsoft.com/en-us/office/dev/add-ins/testing/troubleshoot-manifest#use-runtime-logging-to-debug-your-add-in) to troubleshoot issues with your add-in's XML manifest file, as well as several installation and runtime conditions. Runtime logging writes `console.log` statements to a log file to help you uncover issues. For more information about runtime logging, see [Use runtime logging to debug your add-in](https://docs.microsoft.com/en-us/office/dev/add-ins/testing/troubleshoot-manifest#use-runtime-logging-to-debug-your-add-in).  

Author: ```suggestion
If you are testing your add-in in Office on Windows, you should [enable runtime logging](https://docs.microsoft.com/en-us/office/dev/add-ins/testing/troubleshoot-manifest#use-runtime-logging-to-debug-your-add-in). Runtime logging delivers `console.log` statements to a separate log file you create to help you uncover issues. The statements cover a variety of errors, including errors pertaining to your add-in's XML manifest file, runtime conditions, or installation of your custom functions.  For more information about runtime logging, see [Use runtime logging to debug your add-in](https://docs.microsoft.com/en-us/office/dev/add-ins/testing/troubleshoot-manifest#use-runtime-logging-to-debug-your-add-in).  
```

> ### Check for Excel error messages

Author: Yeah, I see what you're saying. This logs more add-in related errors, rather than end-result errors like #VALUE would. How's the above modification?

> Excel has a number of built-in error messages which are returned to a cell if there is calculation error. These are `#NULL!`, `#DIV/0!`, `#VALUE!`, `#REF!`, `#NAME?`, `#NUM!`, `#N/A`, and `#GETTING_DATA`. At this time, custom functions can only use the error codes listed above when custom functions hit an error.

Author: Makes sense!

## Common issues

> ### Verify SSL certificates

Author: ```suggestion
### My Add-in Will Not Load: Verify Certifications
```

If your add-in fails to register, [verify that the SSL certificates are configured correctly](https://github.com/OfficeDev/generator-office/blob/master/src/docs/ssl.md) for the web server that's hosting your add-in application. Typically if you forget to do this step, you will see an error message in Excel warning you that custom functions could not be installed properly. For more information on this verification, see [Adding self-signed certificates as trusted root certificate](https://github.com/OfficeDev/generator-office/blob/master/src/docs/ssl.md).

> ### Associate your functions

Author: ```suggestion
### My functions won't load: associate functions
```

In your custom functions' script file, you need to both write the function code and code to associate the function's name to the corresponding JSON id. Typically this declaration is made after each function or at the end of the script file. It's common to forget to include this and can be a reason why your functions don't work.

The following example shows an add function, followed by the function's name `add` being associated with the corresponding JSON id `ADD`.

```js
function add(first, second){
  return first + second;
}

CustomFunctions.associate("ADD", add);
```

For more information on this process, see [Associating function names with json metadata](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-best-practices#associating-function-names-with-json-metadata).

### Ensure promises return

> In addition to the ordinary reasons that a cell might report a #GETTING_DATA error, custom functions also will report `#GETTING_DATA` if a promise does not return a result. Check your functions to make sure that any promises are properly returning a result to a cell.

Reviewer: ```suggestion
When Excel is waiting for a custom function to complete, it displays #GETTING_DATA in the cell. If your custom function code returns a promise, but the promise does not return a result, Excel will continue showing #GETTING_DATA. Check your functions to make sure that any promises are properly returning a result to a cell.
```
I figured we should just go ahead and explain why you see GETTING_DATA (but are there other reasons? I just have the one.)

Reviewer: Oh also we have to localize GETTING_DATA and other cell errors, so I left off the tick marks.

Author: There aren't other reasons I know of offhand, but I have seen it occur as a weird problem that crops up (kind of like #VALUE in current build), which is why I'm vague here. However, probably best to assume that it won't crop up if it isn't currently doing so

## Reporting Feedback

> If you are encountering errors that aren't documented here, let us know. The custom functions and documentation teams have several procedures in place for reporting issues.

Reviewer: ```suggestion
If you are encountering issues that aren't documented here, let us know. There are two ways to report issues.
```

### In Excel on Windows or Mac

> If using Excel for Windows or Mac, you can report feedback to the Excel Custom Functions team directly in the UI of Excel itself. To do this, select **File | Feedback | Send a Frown**. Sending a frown will provide the necessary logs to understand the issue you are hitting.

Reviewer: ```suggestion
If using Excel for Windows or Mac, you can report feedback to the Office extensibility team directly from Excel. To do this, select **File -> Feedback -> Send a Frown**. Sending a frown will provide the necessary logs to understand the issue you are hitting.
```

### In Github

> Feel free to submit an issue you encounter either through the "edit" feature on any documentation page, by submitting a [pull request to the office js docs repository](https://github.com/OfficeDev/office-js-docs-pr/blob/master/Contributing.md), or by [filing a new issue directly to the custom functions repository](https://github.com/OfficeDev/Excel-Custom-Functions/issues).

Reviewer: ```suggestion
Feel free to submit an issue you encounter either through the "Content feedback" feature at the bottom of any documentation page, or by [filing a new issue directly to the custom functions repository](https://github.com/OfficeDev/Excel-Custom-Functions/issues).
```
Took out PR instructions because likely they don't know the fix if they are reporting an issue.

Author: Makes sense!
